src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,256 @@
+  struct counters {
+    int64_t ops = 0;
+    int64_t bw = 0;
+  };
+  struct qos_entry {
+    // 2 is the enum size
+    counters read;
+    counters write;
+    ceph::timespan ts;
+    bool first_run = true;
+    std::mutex ts_lock;
+    // Those functions are returning the integer value of the tokens 
+    int64_t read_ops () const
+    {
+      return read.ops / fixed_float_rgw_ratelimit;
+    }
+    int64_t write_ops() const
+    {
+      return write.ops / fixed_float_rgw_ratelimit;
+    }
+    int64_t read_bytes() const
+    {
+      return read.bw / fixed_float_rgw_ratelimit;
+    }
+    int64_t write_bytes() const
+    {
+      return write.bw / fixed_float_rgw_ratelimit;
+    }
+  };
+  static constexpr size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::shared_mutex insert_lock;
+  std::atomic_bool *replacing;
+  std::condition_variable *cv;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  static bool is_read_op(const std::string_view method) {
+    if(method == "GET" || method == "HEAD")
+      return true;
+    return false;
+  }
+
+    // find or create an entry, and return its iterator
+  auto& find_or_create(const std::string& key) {
+    std::shared_lock rlock(insert_lock);
+    if(qos_entries.size() > 0.9 * map_size && *replacing == false)
+    {
+        *replacing = true;
+        cv->notify_all();
+    }
+    auto ret = qos_entries.find(key);
+    rlock.unlock();
+    if(ret == qos_entries.end())
+    {
+      std::unique_lock wlock(insert_lock);
+      ret = qos_entries.emplace(std::piecewise_construct,
+                                 std::forward_as_tuple(key),
+                                 std::forward_as_tuple()).first;
+    }
+    return ret->second;
+  }
+
+  bool should_rate_limit_read(qos_entry& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it.read_ops() - 1 < 0) && (ops_limit > 0)) ||
+      (it.read_bytes() < 0 && bw_limit > 0))
+      return true;
+    // we don't want to reduce ops' tokens if we've rejected it.
+    it.read.ops -= fixed_float_rgw_ratelimit;
+    return false;
+  }
+  /* The purpose of this function is to minimum time before overriding the stored timestamp
+     This function is necessary to force the increase tokens add at least 1 token when it updates the last stored timestamp.
+     That way the user/bucket will not lose tokens because of rounding
+  */
+  bool minimum_time_reached(const ceph::timespan curr_timestamp, const qos_entry& it) const
+  {
+    using namespace std::chrono;
+    constexpr auto min_duration = duration_cast<ceph::timespan>(seconds(60)) / fixed_float_rgw_ratelimit;
+    const auto delta = curr_timestamp - it.ts;
+    if (delta < min_duration)
+      return false;
+    return true;
+  }
+  void increase_tokens(qos_entry& it, const ceph::timespan curr_timestamp,
```

Comment:
```suggestion
  void increase_tokens(qos_entry& it, const ceph::timespan& curr_timestamp,
```

Reply:
As mentioned in the above review:
"ceph::timespan is just a uint64_t, pass it by value"